### ISSUE TITLE
[unified-signatures] Rule is confused by a difference in signatures on an async function's return type

### ISSUE THREAD

**Issue by mdouglass** [typescript-eslint/typescript-eslint#740]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/unified-signature": "error"
  }
}
```

```TS
function get1(key: string): string | undefined
function get1(key: string, defaultValue: string): string
function get1(key: string, defaultValue?: string): string | undefined
{
  const obj: Record<string, string> = { }
  return obj[key] || defaultValue
}

function get2(key: string): Promise<string | undefined>
function get2(key: string, defaultValue: string): Promise<string> // <-- warning here
async function get2(key: string, defaultValue?: string): Promise<string | undefined>
{
  const obj: Record<string, string> = { }
  return obj[key] || defaultValue
}
```

**Expected Result**
Neither get1 nor get2 have any warnings

**Actual Result**
get2 issues a unified signature warning on `defaultValue: string` in line 2 of get2's declaration. I don't see a way to combine the signatures without losing the information that the return type differs between the two overloads.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.12.0` |
| `@typescript-eslint/parser`        | `1.12.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `6.1.0` |
| `node`                             | `12.6.0` |
| `npm`                              | `6.9.0` |

**Comment by komcal** [comment]
Can I fix this pr? @bradzacher

**Comment by bradzacher** [comment]
You don't need to ask, @komcal - anyone is welcome to submit PR for issues without the "has pr" tag.
These issues with "good first issue" are intended to be nice ones for new contributors to get stuck into, so feel free!
